x-pack/platform/plugins/shared/onechat/server/services/tools/builtin/definitions/create_visualization/create_visualization.ts

```diff
@@ -0,0 +1,146 @@
+import type { BuiltinToolDefinition } from '@kbn/onechat-server';
+import { ToolResultType } from '@kbn/onechat-common/tools/tool_result';
+import parse from 'joi-to-json';
+
+import { esqlMetricState } from '@kbn/lens-embeddable-utils/config_builder/schema/charts/metric';
+import { getToolResultId } from '@kbn/onechat-server';
+import { getChartType } from './guess_chart_type';
+import { SupportedChartType } from './types';
+import { createVisualizationGraph } from './graph_lens';
+
+const metricSchema = parse(esqlMetricState.getSchema()) as object;
+
+const createVisualizationSchema = z.object({
+  query: z.string().describe('A natural language query describing the desired visualization.'),
+  existingConfig: z
+    .string()
+    .optional()
+    .describe('An existing visualization configuration to modify.'),
+  chartType: z
+    .enum([SupportedChartType.Metric, SupportedChartType.Map])
+    .optional()
+    .describe(
+      '(optional) The type of chart to create as indicated by the user. If not provided, the LLM will suggest the best chart type.'
+    ),
+  esql: z
+    .string()
+    .optional()
+    .describe(
+      '(optional) An ES|QL query. If not provided, tool with automatically generate the query.'
+    ),
+});
+
+export const createVisualizationTool = (): BuiltinToolDefinition<
+  typeof createVisualizationSchema
+> => {
+  return {
+    id: platformCoreTools.createVisualization,
+    type: ToolType.builtin,
+    description: `Create a visualization configuration based on a natural language description.
+
+This tool will:
+1. Determine the best chart type if not specified (from: ${Object.values(SupportedChartType).join(
+      ', '
+    )})
+2. Generate an ES|QL query if not provided
+3. Generate a valid visualization configuration`,
+    schema: createVisualizationSchema,
+    tags: [],
+    handler: async (
+      { query: nlQuery, chartType, esql, existingConfig },
+      { esClient, modelProvider, logger, events }
+    ) => {
+      try {
+        // Step 1: Determine chart type if not provided
+        let selectedChartType: SupportedChartType = chartType || SupportedChartType.Metric;
+        const parsedExistingConfig = existingConfig ? JSON.parse(existingConfig) : null;
+
+        if (!chartType) {
+          logger.debug('Chart type not provided, using LLM to suggest one');
+          const suggestedChartType = await getChartType(
+            modelProvider,
+            parsedExistingConfig?.type,
+            nlQuery
+          );
+          if (
+            Object.values(SupportedChartType).includes(suggestedChartType as SupportedChartType)
+          ) {
+            selectedChartType = suggestedChartType as SupportedChartType;
+          } else {
+            logger.warn(
+              `LLM suggested invalid chart type: ${suggestedChartType}, defaulting to metric`
+            );
+            selectedChartType = SupportedChartType.Metric;
+          }
+        }
```

Comment:
NIT: Should this part move inside `getChartType`? feel like the function should always return a valid `SupportedChartType` or throw, we shouldn't have to do that additional check , no? If we want to control the behavior, we could add a `defaultChartType` optional param to the function, defaulting to `Metric`?